src/out_ref.rs

```diff
@@ -0,0 +1,573 @@
+                    "being present in debug mode).\n",
+                    r"/!\ THIS IS A BUG AND A SOUNDNESS ISSUE /!\", "\n",
+                    "Please submit an issue ASAP.",
+                ));
+            } else {
+                ::core::hint::unreachable_unchecked()
+            })
+    }
+
+    #[inline]
+    pub
+    unsafe
+    fn assume_init (mut self: Out<'out, [T]>) -> &'out mut [T]
+    {
+        let len = self.len();
+        slice::from_raw_parts_mut(
+            self.as_mut_ptr(),
+            len,
+        )
+    }
+
+    /// Initialize the buffer with a copy from another (already initialized)
+    /// buffer.
+    ///
+    /// It returns a read-writable slice to the initialized bytes for
+    /// convenience (automatically
+    /// [`assume_init`][`Out::assume_init`]-ed).
+    ///
+    /// # Panic
+    ///
+    /// The function panics if the slices' lengths are not equal.
+    ///
+    /// # Guarantees
+    ///
+    /// A non-`panic!`king return from this function **guarantees that the input
+    /// slice has been (successfully) initialized**, and that it is thus then
+    /// sound to `.assume_init()`.
+    ///
+    /// It also guarantees that the returned slice does correspond to the input
+    /// slice (_e.g._, for [`crate::ReadIntoUninit`]'s safety guarantees).
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use ::core::mem::{self, MaybeUninit};
+    /// use ::uninit::prelude::*;
+    ///
+    /// let mut array = uninit_array![_; 13];
+    /// assert_eq!(
+    ///     array.as_out().copy_from_slice(b"Hello, World!"),
+    ///     b"Hello, World!",
+    /// );
+    /// // we can thus soundly `assume_init` our array:
+    /// let array = unsafe {
+    ///     mem::transmute::<
+    ///         [MaybeUninit<u8>; 13],
+    ///         [            u8 ; 13],
+    ///     >(array)
+    /// };
+    /// assert_eq!(
+    ///     array,
+    ///     *b"Hello, World!",
+    /// );
+    /// ```
+    pub
+    fn copy_from_slice (
+        mut self: Out<'out, [T]>,
+        source_slice: &'_ [T],
+    ) -> &'out mut [T]
+    where
+        T : Copy,
+    {
+        unsafe {
+            // # Safety
+            //
+            //   - Writing to `self.0` is fine since `source_slice` only
+            //     contains initialized elements.
+            //
+            //   - the `copy_nonoverlapping()` call guarantees that the buffer
+            //     has been initialized.
+            self.r()
+                .as_uninit_mut()
+                .copy_from_slice(
+                    <[MaybeUninit<T>]>::from_ref(source_slice)
+                )
+            ;
+            self.assume_init()
+        }
+    }
+
+    /// Fill the buffer with values from a function mapping indices to values.
+    ///
+    /// # Guarantees
+    ///
+    /// `factory` is guaranteed to be called in the `0 .. self.len()` order.
+    #[inline]
+    pub
+    fn fill_with (
```

Comment:
I get where this is coming from. That said, no other initialization interface uses `FnMut(usize)` and if even the guarantee of iteration order is provided and not a generic `Fn` is used then the index argument is more or less redundant. 

Somehow, I would also have thought of an `impl IntoIterator<Item=T>` argument instead of a function and a return of the prefix that the iterator was able to provide. This is strictly more powerful due to the relaxed precondition while being able to implement this method by using the argument `(..).map(factory)`.